rpc/src/optimistically_confirmed_bank_tracker.rs

```diff
@@ -191,6 +195,29 @@ impl OptimisticallyConfirmedBankTracker {
         }
     }
 
+    fn notify_root_banks(
+        bank: &Arc<Bank>,
+        slot_threshold: Slot,
+        bank_notification_subscribers: &Option<Arc<RwLock<Vec<BankNotificationSender>>>>,
+    ) {
+        if bank_notification_subscribers.is_none() {
+            return;
+        }
+        for root_bank in bank.clone().parents_inclusive().iter().rev() {
```

Comment:
i'm not sure `Bank::parents_inclusive()` is what we want to use here. it's going to be a slow linked-list walk to build the `Vec` that it returns and iirc, what's available is limited to what's in the local node's ledger.  `Bank::ancestors` is probably what we want instead.

i think we should also `take_while` from current root back to the last reported root, reverse that, then notify forward, to reduce considering very old roots that have almost certainly been notified already

Reply:
Thanks! The code in line 180 uses the same logic. Should we do the same optimization?

Reply:
Unfortunately -- we cannot use Bank::ancestors -- the notification interfaces requires a Bank for the ancestor

Reply:
Do we have any assurances that the ancestor roots haven't already been dropped from BankForks?

Reply:
Great question. Looking into it!

Reply:
`ancestors` is not safe to use because there could have been new roots made by the time this is called since this is async from replay_stage where the roots are set.

What we probably want is to send a `Vec<Arc<Bank>>` from `replay_stage` where the root is being set. In `set_root` We are already building a vector of parents of the current root, which should include all the parents up until the last root https://github.com/AshwinSekar/solana/blob/2efd72b1a1a06a89613232d92f75a8e1c1995a6a/core/src/replay_stage.rs#L2023-L2024. We could just grab those and send them over this channel.

A more detailed questions though is why do we need the actual banks in these notifications? I'm assuming to access AccountsDB? But having the bank `N` does not guarantee AccountsDB is in the same view as it was during bank `N` because future roots `> N` could have been made and cleaned up state

Reply:
> the notification interfaces requires a Bank for the ancestor

doesn't look like this is true? i think we only pull the slot and parent slot out of the bank when forwarding the notification. so could extract those earlier and only forward them along?

Reply:
> To Tyera's point `ancestors()` is not safe to use because there could have been new roots made by the time this is called since this is async from `replay_stage` where the roots are set, and then earlier rooted banks would have been removed from `BankForks`
> 
> What we probably want is to send a `Vec<Arc<Bank>>` from `replay_stage` where the root is being set. In `set_root` We are already building a vector of parents of the current root, which should include all the parents up until the last root https://github.com/AshwinSekar/solana/blob/2efd72b1a1a06a89613232d92f75a8e1c1995a6a/core/src/replay_stage.rs#L2023-L2024. We could just grab those and send them over this channel.
> 
> A more detailed questions though is why do we need the actual banks in these notifications? I'm assuming to access AccountsDB? But having the bank `N` does not guarantee AccountsDB is in the same view as it was during bank `N` because future roots `> N` could have been made and cleaned up state

Thanks @carllin -- as pointed by @t-nelson, we can change the notifications by sending the (slot, parent_slot)  pair without requiring the bank. Regarding the race condition on the roots on the BankForks -- I wonder why https://github.com/AshwinSekar/solana/blob/2efd72b1a1a06a89613232d92f75a8e1c1995a6a/rpc/src/optimistically_confirmed_bank_tracker.rs#L176 does not have the same concern?

Reply:
I think the same issue is present there as well :)

Reply:
i guess we could reduce the scope of this pr to "be consistently broken", rather than uniquely broken? :slightly_smiling_face:

Reply:
Or fix the other place first...?

Reply:
> Or fix the other place first...?

I will fix this issue first and the investigate the other next to reduce the scope.